File: transactions/src/main/java/com/example/repository/AccountRepository.java
package com.example.repository;

import com.example.model.Account;
import jakarta.persistence.LockModeType;
import jakarta.persistence.QueryHint;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.jpa.repository.QueryHints;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface AccountRepository extends JpaRepository<Account, Long> {
    List<Account> findByUserId(String userId);
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT a FROM Account a WHERE a.id = :accountId")
    @QueryHints({
            @QueryHint(name = "jakarta.persistence.lock.timeout", value = "5000")
    })
    Optional<Account> findByIdForUpdate(@Param("accountId") Long accountId);
}

File: transactions/src/main/kotlin/com/example/repository/AccountRepository.kt
package com.example.repository

import com.example.model.Account
import org.springframework.data.r2dbc.repository.Query
import org.springframework.data.repository.kotlin.CoroutineCrudRepository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

interface AccountRepository : CoroutineCrudRepository<Account, Long> {
    @Query("SELECT id, user_id, currency FROM account WHERE user_id = :userId")
    fun findByUserId(userId: String): Flux<Account>

    @Query("SELECT id, user_id, currency FROM account WHERE id = :accountId")
    fun findByIdForUpdate(accountId: Long): Mono<Account>
}